src/h3lib/lib/polygon.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2018 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/** @file polygon.c
+ * @brief Polygon algorithms
+ */
+
+#include "polygon.h"
+#include <float.h>
+#include <stdbool.h>
+#include "bbox.h"
+#include "geoCoord.h"
+#include "h3api.h"
+
+/**
+ * Normalize a longitude value, converting it into a comparable value in
+ * transmeridian cases.
+ * @param  lng Longitude to normalize
+ * @param  isTransmeridian Whether this longitude is part of a transmeridian
+ *                         polygon
+ * @return Normalized longitude
+ */
+double _normalizeLng(double lng, bool isTransmeridian) {
+    return isTransmeridian && lng < 0 ? lng + M_2PI : lng;
+}
+
+/**
+ * loopContainsPoint is the core loop of the point-in-poly
+ * algorithm, working on a Geofence struct
+ *
+ * @param geofence The geofence to check
+ * @param bbox The bbox for the loop being tested
+ * @param coord The coordinate to check if contained by the geofence
+ * @return true or false
+ */
+bool loopContainsPoint(const Geofence* geofence, const BBox* bbox,
+                       const GeoCoord* coord) {
+    // fail fast if we're outside the bounding box
+    if (!bboxContains(bbox, coord)) {
+        return false;
+    }
+    bool isTransmeridian = bboxIsTransmeridian(bbox);
+    bool contains = false;
+
+    double lat = coord->lat;
+    double lng = _normalizeLng(coord->lon, isTransmeridian);
+
+    for (int i = 0; i < geofence->numVerts; i++) {
```

Comment:
Could you refactor the body of the loop into a helper like `static GeoCoord pairContainsPoint(const GeoCoord* a, const GeoCoord* b);`? That would allow iteration logic to lie in caller but ray casting algo to be generic.